pkg/applicationserver/io/grpc/grpc.go

```diff
@@ -20,20 +20,27 @@ import (
 	pbtypes "github.com/gogo/protobuf/types"
 	"go.thethings.network/lorawan-stack/pkg/applicationserver/io"
 	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/config"
 	"go.thethings.network/lorawan-stack/pkg/errors"
 	"go.thethings.network/lorawan-stack/pkg/log"
 	"go.thethings.network/lorawan-stack/pkg/ttnpb"
 	"go.thethings.network/lorawan-stack/pkg/unique"
 	"google.golang.org/grpc/peer"
 )
 
+// MQTTConfigProvider provides contextual access to MQTT configuration.
+type MQTTConfigProvider interface {
+	GetMQTTConfig(context.Context) (*config.MQTT, error)
+}
+
 type impl struct {
-	server io.Server
+	server   io.Server
+	provider MQTTConfigProvider
 }
 
 // New returns a new gRPC frontend.
-func New(server io.Server) ttnpb.AppAsServer {
-	return &impl{server}
+func New(server io.Server, provider MQTTConfigProvider) ttnpb.AppAsServer {
```

Comment:
I'm afraid that adding args to the constructor isn't the best way to go as we'll likely keep adding more options. 

How about using variadic options?

```go
type Option interface {
	apply(*impl)
}

type optionFunc func(*impl)

func (f optionFunc) apply(i *impl) { f(i) }

func WithMQTTConfigProvider(provider MQTTConfigProvider) Option {
	return optionFunc(func(i *impl) {
		i.mqttConfigProvider = provider
	})
}

func New(server io.Server, opts ...Option) ttnpb.GtwGsServer {
	i := &impl{server}
	for _, opt := range opts {
		opt.apply(i)
	}
	return i
}
```